sale_coupon_criteria_order_based/models/coupon_program.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2022 Ooops404
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import ast
+
+from odoo import _, api, fields, models
+
+
+class CouponProgram(models.Model):
+    _inherit = "coupon.program"
+
+    rule_order_domain = fields.Char(
+        string="Based on Order",
+        help="Coupon program will work for the order with selected domain only",
+    )
+
+    def _is_valid_order(self, order):
+        """
+        Check that we can apply the coupon for current order
+        """
+        if self.rule_order_domain and self.rule_order_domain != "[]":
+            domain = ast.literal_eval(self.rule_order_domain) + [("id", "=", order.id)]
+            return bool(self.env["sale.order"].search_count(domain))
```

Comment:
as you already have the order and you are filtering by id just apply the domain on it
```suggestion
            return bool(order.filtered_domain(domain))
```